pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -283,39 +284,68 @@ func balanceDomains(
+				klog.V(2).InfoS(err.Error(), "pod", klog.KObj(aboveToEvict[k]))
 				continue
 			}
+
 			podsForEviction[aboveToEvict[k]] = struct{}{}
 		}
 		sortedDomains[j].pods = sortedDomains[j].pods[:len(sortedDomains[j].pods)-movePods]
 		sortedDomains[i].pods = append(sortedDomains[i].pods, aboveToEvict...)
 	}
 }
 
-// nodesPodFitsOnBesidesCurrent counts the number of nodes this pod could fit on based on its affinity
+// validatePodFitsOnOtherNodes performs validation based on scheduling predicates for affinity and toleration.
 // It excludes the current node because, for the sake of domain balancing only, we care about if there is any other
 // place it could theoretically fit.
 // If the pod doesn't fit on its current node, that is a job for RemovePodsViolatingNodeAffinity, and irrelevant to Topology Spreading
-func nodesPodFitsOnBesidesCurrent(pod *v1.Pod, nodeMap map[string]*v1.Node) int {
-	count := 0
+func validatePodFitsOnOtherNodes(pod *v1.Pod, nodeMap map[string]*v1.Node) error {
+	// if the pod has a hard nodeAffinity/nodeSelector/toleration that only matches this node,
+	// don't bother evicting it as it will just end up back on the same node
+	// however we still account for it "being evicted" so the algorithm can complete
+	// TODO(@damemi): Since we don't order pods wrt their affinities, we should refactor this to skip the current pod
+	// but still try to get the required # of movePods (instead of just chopping that value off the slice above)
+	isRequiredDuringSchedulingIgnoredDuringExecution := pod.Spec.Affinity != nil &&
+		pod.Spec.Affinity.NodeAffinity != nil &&
+		pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil
+
+	hardTaintsFilter := func(taint *v1.Taint) bool {
+		return taint.Effect == v1.TaintEffectNoSchedule || taint.Effect == v1.TaintEffectNoExecute
+	}
+
+	var availableTargetNodes, ineligibleAffinityCount, taintNotToleratedCount int
 	for _, node := range nodeMap {
-		if nodeutil.PodFitsCurrentNode(pod, node) && node != nodeMap[pod.Spec.NodeName] {
-			count++
+		if node == nodeMap[pod.Spec.NodeName] {
+			continue
+		}
+
+		if pod.Spec.NodeSelector != nil || isRequiredDuringSchedulingIgnoredDuringExecution {
+			if !nodeutil.PodFitsCurrentNode(pod, node) {
+				ineligibleAffinityCount++
+				continue
+			}
+		}
+
+		if !utils.TolerationsTolerateTaintsWithFilter(pod.Spec.Tolerations, node.Spec.Taints, hardTaintsFilter) {
+			taintNotToleratedCount++
+			continue
+		}
+
+		availableTargetNodes++
+	}
+
+	if availableTargetNodes == 0 {
+		var errs []error
+		if ineligibleAffinityCount > 0 {
+			errs = append(errs, fmt.Errorf("ignoring pod for eviction due to %d nodes with ineligible selector/affinity", ineligibleAffinityCount))
+		}
+		if taintNotToleratedCount > 0 {
+			errs = append(errs, fmt.Errorf("ignoring pod for eviction due to %d nodes with taints that are not tolerated", taintNotToleratedCount))
 		}
+		return utilerrors.NewAggregate(errs)
```

Comment:
Thanks, this looks a little better. Though, I more had in mind aggregating the errors within the node loop (but I wasn't clear about that, sorry).

But I like this direction. Instead of just incrementing a count, could you build a slice of the actual node names that don't match? Then change each of these errors to list the names, like:
```go
errs = append(errs, fmt.Errorf("%d nodes with ineligible selector/affinity: %+v", len(ineligibleAffinityNodes), ineligibleAffinityNodes))
```

and in the call to this function, prepend the error info there:
```go
klog.V(2).InfoS(fmt.Errorf("ignoring pod for eviction due to: %s", err.Error()), "pod", klog.KObj(aboveToEvict[k]))
```

Though, listing the actual node names might be a bit too much on large clusters. So I'm okay with not doing that, however I would like to remove the duplicate text. After that lgtm